File: mixed/f1/group1/PathSum.java
// Given the root of a binary tree and an integer targetSum, return true if the tree has a root-to-leaf path
// such that adding up all the values along the path equals targetSum.
// A leaf is a node with no children.
// Sample 1:
// Input: root = [5,4,8,11,null,13,4,7,2,null,null,null,1], targetSum = 22
// Output: true
// Explanation: The root-to-leaf path with the target sum has leaf node 2.
// Note: visualize to understand

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}
class PathSum {
    public static void main(String...args) {
        TreeNode tree = new TreeNode(5);
        TreeNode left = new TreeNode(4);
        tree.left = left;
        TreeNode leftLeft = new TreeNode(11, new TreeNode(7), new TreeNode(2));
        tree.left.left = leftLeft;

        TreeNode right = new TreeNode(8);
        TreeNode rightLeft = new TreeNode(13);
        TreeNode rightRight = new TreeNode(4, null, new TreeNode(1));
        right.left = rightLeft;
        right.right = rightRight;
        tree.right = right;

        int targetSum = 22;
        System.out.println("Path sum exists? " + hasPathSum(tree, targetSum));

    }

    static boolean hasPathSum(TreeNode root, int targetSum) {
        int sum = 0;
        return rootToLeafPathSum(root, targetSum, sum);
    }
    private static boolean rootToLeafPathSum(TreeNode root, int targetSum, int sum) {
        if (root == null)
            return false;
        if (root.left == null && root.right == null) {
            sum = sum + root.val;
            if (sum == targetSum)
                return true;
        }

        return rootToLeafPathSum(root.left, targetSum, sum + root.val)
                || rootToLeafPathSum(root.right, targetSum, sum + root.val);
    }
}

File: bytopic/binarytree/python/group1/path_sum.py
"""
Given the root of a binary tree and an integer targetSum, return true if the tree has a root-to-leaf path
such that adding up all the values along the path equals targetSum.
A leaf is a node with no children.

Sample 1:
Input: root = [5,4,8,11,null,13,4,7,2,null,null,null,1], targetSum = 22
Output: true
Explanation: The root-to-leaf path with the target sum has leaf node 2.
Note: visualize to understand

Tag: 76/150
Tag: 112/2927, R966/2936 (overall frequency ranking)
"""

from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def has_path_sum(root: Optional[TreeNode], target_sum: int) -> bool:
    if not root:
        return False
    if not root.left and not root.right:
        return target_sum == root.val
    left_sum = has_path_sum(root.left, target_sum - root.val)
    right_sum = has_path_sum(root.right, target_sum - root.val)
    return left_sum or right_sum


def main():
    root = TreeNode(5, TreeNode(4), TreeNode(8))
    root.left.left = TreeNode(11)
    root.left.left.left = TreeNode(7)
    root.left.left.right = TreeNode(2)
    root.right.left = TreeNode(13)
    root.right.right = TreeNode(4)
    root.right.right.right = TreeNode(1)

    target_sum = 22
    print(has_path_sum(root, target_sum))


if __name__ == "__main__":
    main()
